File: Data Structures & Algorithms/max-water-container/submission-2.java
class Solution {
    public int maxArea(int[] heights) {
        
        int maxArea = 0;
        int l = 0, r = heights.length - 1;

        while(l < r)
        {
            int h = Math.min(heights[l],heights[r]);
            int w = r-l;
            int area = h * w;

            maxArea = Math.max(maxArea,area);

            if(heights[l] < heights[r])
                l++;
            else
                r--;
        }

        return maxArea;
    }
}

File: Data Structures & Algorithms/max-area-of-island/submission-2.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        n = len(grid)
        m = len(grid[0])
        ans = 0

        def dfs(i,j):
            if i < 0 or i == n or j < 0 or j == m or grid[i][j] == 0:
                return
            #Consume the cell value
            grid[i][j] = 0
            #Increment the area value
            self.area += 1
            dfs(i-1,j)
            dfs(i+1,j)
            dfs(i,j-1)
            dfs(i,j+1)

        for i in range(n):
            for j in range(m):
                if grid[i][j] == 1:
                    #Reset the area as its start of a new island
                    self.area = 0
                    dfs(i,j)
                    ans = max(ans,self.area)
        
        return ans